File: SDK Examples/Examples/Matlab/Counter_BufferedUpDownCounter/BufferedUpDownCounter.m
% BufferedUpDownCounter.m
%
% Example Category:
%    Counter
% Matlab(2010 or 2010 above)
%
% Description:
%    This example demonstrates how to use BufferedUpDown Counter function.
%
% Instructions for Running:
%    1. Set the 'deviceDescription' for opening the device. 
%    2. Set the 'chEnable' to select channel of the counter to operate.
%    3. Set the 'channelCount' to decide how many channels to operate. 
%    4. Set the 'sectionLength'to decide what occasion to signal event;
%    5. Set the 'sectionCount' to decide the count of data section to 
%       acquire,the operation will stop automatically after the specified 
%       count of data section is got.
%    6. Set the 'sampleClkSrc' to set the sampling clock source.
%
% I/O Connections Overview:
%    Please refer to your hardware reference manual.

function BufferedUpDownCounter()

% Make Automation.BDaq assembly visible to MATLAB.
BDaq = NET.addAssembly('Automation.BDaq4');

% Configure the following parameters before running the demo.
% The default device of project is PCIE-1813, users can choose other 
% devices according to their needs. 
deviceDescription = 'PCIE-1813,BID#15';  
channelCount = int32(4);
chEnable = [1,0,0,0]; 
sectionLength = int32(1024); 
sectionCount = int32(0); % "sectionCount" is zero, meaning Streaming Mode.
sampleClkSrc = Automation.BDaq.SignalDrop.SigExtDigClock;

% triggerAction = Automation.BDaq.TriggerAction.DelayToStart;
triggerSource = Automation.BDaq.SignalDrop.SignalNone;
% triggerEdge = Automation.BDaq.ActiveSignal.RisingEdge;
% triggerLevel = double(3.0);

% Step 1: Create a 'bfdUdCounterCtrl' for BufferedUpDown Counter function.
bfdUdCounterCtrl = Automation.BDaq.BufferedUdCounterCtrl();

% Step 2: Set the notification event Handler by which we can known the 
% state of operation effectively. 
addlistener(bfdUdCounterCtrl, 'DataReady', @bfdUdCounterCtrl_DataReady);

try
    % Step 3: Select a device by device number or device description and 
    % specify the access mode. In this example we use 
    % ModeWrite(default) mode so that we can fully control the 
    % device, including configuring, sampling, etc.
    bfdUdCounterCtrl.SelectedDevice = Automation.BDaq.DeviceInformation(...
        deviceDescription);
   
    if 	bfdUdCounterCtrl.Features.TriggerCount
    else
        e = MException('DAQError:NotSupport', ...
            'The device do not support trigger function!');
        throw (e);
    end
    
    % Step 4: Set necessary parameters for counter operation.
    bfdUdCounterCtrl.ChannelMap = chEnable;
    
    record = bfdUdCounterCtrl.Record;
    record.SectionLength = sectionLength;
    record.SectionCount = sectionCount;
   
    bfdUdCounterCtrl.SampleClkSrc = sampleClkSrc;
    
    % Setp 5: Trigger parameters setting.
    % Diffenert device with different trigger source, when changing trigger
    % source refer to manual for details.
    trigger = bfdUdCounterCtrl.Trigger;
    if ~isempty(trigger)
        trigger.Source = triggerSource;
        %trigger.Action = triggerAction;
        %trigger.Edge = triggerEdge;
        %trigger.Level = triggerLevel;
    else
        e = MException('DAQError:NotSupport',...
            'The device do not support trigger function!');
        throw (e);
    end
    
    % Step 6: Set counting type for BufferedUpDown Counter.
    % counting type : CountingNone,DownCount,UpCount,PulseDirection,
    % TwoPulse,AbPhaseX1,AbPhaseX2,AbPhaseX4.
    signalCountingType = Automation.BDaq.CountingType.PulseDirection;
    for i = 1 : (channelCount)
        if(chEnable(i))
            getChannels = bfdUdCounterCtrl.Channels.Get(i);
            getChannels.CountingType = signalCountingType;
        end    
    end 

    % Step 7: Start BufferedUpDown Counter. 
    errorCode = bfdUdCounterCtrl.Prepare();
    if BioFailed(errorCode)
        throw Exception(); 
    end
    bfdUdCounterCtrl.Enabled = true;
    
    % Step 8: Read counting value: connect the input signal to channels
    % you selected to get event counter value.
    fprintf('Buffered UpDown Counter is in progress...\n');
    fprintf('Connect the input signal to the connector.\n');
    input('Press Enter key to quit!\n\n','s');
 
    % Step 9: Stop BufferedUpDown Counter.
    bfdUdCounterCtrl.Enabled = false;
catch e
    % Something is wrong. 
    errStr = e.message;
    disp(errStr);
end   

% Step 10: Close device and release any allocated resource.
bfdUdCounterCtrl.Dispose();
end

function result = BioFailed(errorCode)
result =  errorCode < Automation.BDaq.ErrorCode.Success && ...
    errorCode >= Automation.BDaq.ErrorCode.ErrorHandleNotValid;
end

function bfdUdCounterCtrl_DataReady(sender, e)
bfdUdCounterCtrl = sender;
% buffer section length, when 'DataReady' event been signaled,
% driver renew data count is e.count. 
   dataBuf = NET.createArray('System.Int32', e.Count);
   [errorCode, returnedCount, startTime] = bfdUdCounterCtrl.GetData(e.Channel, e.Count, dataBuf, 0);
   fprintf('BufferedUpDownCounter DataReadyEvent, channel %d Current Buffered UpDown counter counts  :%d\n\n', ...
        e.Channel, dataBuf(1));
end

File: SDK Examples/Examples/ANSI_C/Counter_BufferedUpDownCounter/BufferedUpDownCounter.c
/*******************************************************************************
Copyright (c) 1983-2017 Advantech Co., Ltd.
********************************************************************************
THIS IS AN UNPUBLISHED WORK CONTAINING CONFIDENTIAL AND PROPRIETARY INFORMATION
WHICH IS THE PROPERTY OF ADVANTECH CORP., ANY DISCLOSURE, USE, OR REPRODUCTION,
WITHOUT WRITTEN AUTHORIZATION FROM ADVANTECH CORP., IS STRICTLY PROHIBITED. 
================================================================================
REVISION HISTORY
--------------------------------------------------------------------------------
$Log: $
--------------------------------------------------------------------------------
$NoKeywords:  $
*/
/******************************************************************************
*
* Windows Example:
*     BufferedUpDownCounter.c
*
* Example Category:
*    Counter
*
* Description:
*    This example demonstrates how to use Buffered UpDown Counter function.
*
* Instructions for Running:
*    1  Set the 'deviceDescription' for opening the device. 
*	  2  Set the 'profilePath' to save the profile path of being initialized device. 
*
* I/O Connections Overview:
*    Please refer to your hardware reference manual.
*
******************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include "../inc/compatibility.h"
#include "../../../inc/bdaqctrl.h"
//-----------------------------------------------------------------------------------
// Configure the following parameters before running the demo
//-----------------------------------------------------------------------------------
#define     deviceDescription L"PCIE-1813,BID#15"
const wchar_t* profilePath = L"../../profile/PCIE-1813.xml";
#define     sectionLength 1024
#define     sectionCount  0

// user buffer size should be equal or greater than raw data buffer length, because data ready count
// is equal or more than smallest section of raw data buffer and up to raw data buffer length.
// users can set 'USER_BUFFER_SIZE' according to demand.
#define    USER_BUFFER_SIZE   sectionLength
int32      userDataBuffer[USER_BUFFER_SIZE];

void waitAnyKey()
{
   do{SLEEP(1);} while(!kbhit());
} 

void BDAQCALL OnDataReadyEvent(void * sender, BfdCntrEventArgs * args, void *userParam)
{	
	int32 returnedCount = 0;
   ErrorCode ret = Success;
	BufferedUdCounterCtrl *ctrl = (BufferedUdCounterCtrl *)sender;
   int32 getDataCount = min(USER_BUFFER_SIZE, args->Count);
	ret = BufferedUdCounterCtrl_GetDataI32(ctrl, args->Channel, getDataCount, userDataBuffer, 0, &returnedCount, NULL);
	// in this demo, we show only the first sample of each channel's new data
   if(!BioFailed(ret)) {
      printf("Buffered Up-down Counting get data count is  %d\n", returnedCount);
      printf("the first sample for Channel %d is: %d\n", args->Channel, userDataBuffer[0]);
   }
}
// This function is used to deal with 'Overrun' Event.
void BDAQCALL OnOverRunEvent(void * sender, BfdCntrEventArgs * args, void *userParam)
{
	printf("Buffered Up-down Counting Overrun: channel = %d, offset = %d, count = %d\n", args->Channel, args->Offset, args->Count);
}
// This function is used to deal with 'CacheOverflow' Event.
void BDAQCALL OnCacheOverflowEvent(void * sender, BfdCntrEventArgs * args, void *userParam)
{
	printf("Buffered Up-down Counting Cache Overflow: channel = %d, offset = %d, count = %d\n", args->Channel, args->Offset, args->Count);
}
// This function is used to deal with 'Stopped' Event.
void BDAQCALL OnStoppedEvent(void * sender, BfdCntrEventArgs * args, void *userParam)
{
   printf("Buffered Up-down Counting stopped: channel = %d, offset = %d, count = %d\n", args->Channel, args->Offset, args->Count);
}

int main(int argc, char* argv[])
{
   ErrorCode ret     = Success;
   IArray* channels  = NULL;
   int32 chCount     = 4;
   int8  chEnabled[] = {1, 0, 0, 0};
   Trigger*  trigger = NULL;
   Record*   record  = NULL;
   UdChannel * udChannel = NULL;
   int i = 0;

   // Step 1: Create a 'BufferedUdCounterCtrl' for Buffered UpDown Counter function.
   BufferedUdCounterCtrl* bfdUdCounterCtrl = BufferedUdCounterCtrl_Create();

   // Step 2: Set the notification event Handler by which we can known the state of operation effectively.
   BufferedUdCounterCtrl_addDataReadyHandler(bfdUdCounterCtrl, OnDataReadyEvent, NULL);
   BufferedUdCounterCtrl_addOverrunHandler(bfdUdCounterCtrl, OnOverRunEvent, NULL);
   BufferedUdCounterCtrl_addCacheOverflowHandler(bfdUdCounterCtrl, OnCacheOverflowEvent, NULL);
   BufferedUdCounterCtrl_addStoppedHandler(bfdUdCounterCtrl, OnStoppedEvent, NULL);
   do
   {
      // Step 3: Select a device by device number or device description and specify the access mode.
      // in this example we use ModeWrite mode so that we can fully control the device, including configuring, sampling, etc.
      DeviceInformation devInfo;
      devInfo.DeviceNumber = -1;
      devInfo.DeviceMode   = ModeWrite;
      devInfo.ModuleIndex  = 0;
      wcscpy(devInfo.Description, deviceDescription);
      ret = BufferedUdCounterCtrl_setSelectedDevice(bfdUdCounterCtrl, &devInfo);
      CHK_RESULT(ret);
      ret = BufferedUdCounterCtrl_LoadProfile(bfdUdCounterCtrl, profilePath);//Loads a profile to initialize the device.
      CHK_RESULT(ret);

      // Step 4: Set necessary parameters.
      ret = BufferedUdCounterCtrl_setChannelMap(bfdUdCounterCtrl, chCount, chEnabled);
	   CHK_RESULT(ret);

      ret = BufferedUdCounterCtrl_setSampleClkSrc(bfdUdCounterCtrl, SigExtDigClock);
      CHK_RESULT(ret);

      record = BufferedUdCounterCtrl_getRecord(bfdUdCounterCtrl);
      {
         ret = Record_setSectionCount(record, sectionCount); //The 0 means setting 'streaming' mode
         CHK_RESULT(ret);
         ret = Record_setSectionLength(record, sectionLength);
         CHK_RESULT(ret);
      }

      trigger = BufferedUdCounterCtrl_getTrigger(bfdUdCounterCtrl); //Setting trigger
      if(trigger) {
         ret = Trigger_setSource(trigger, SignalNone); // The "SignalNone" means doing without trigger.
         CHK_RESULT(ret);
      }

      // Step 5: Set counting type for  Buffered UpDown Counter
		/******************************************************************************************************************/
		/*In this example, we use the PCIE-1813 and set 'PulseDirection' as the default CountingType.See manual for details.
		/******************************************************************************************************************/
      channels = BufferedUdCounterCtrl_getChannels(bfdUdCounterCtrl);
      for (i = 0; i < chCount; i++)
      {
         udChannel = (UdChannel*)Array_getItem(channels, i);
         ret = UdChannel_setCountingType(udChannel, PulseDirection);
         CHK_RESULT(ret);
      }
      // Step 6: Start Buffered UpDown Counter 
      ret = BufferedUdCounterCtrl_Prepare(bfdUdCounterCtrl);
      CHK_RESULT(ret);

	   ret= BufferedUdCounterCtrl_setEnabled(bfdUdCounterCtrl, TRUE);
	   CHK_RESULT(ret);

	   // Step 7: Read counting value: connect the input signal to channels you selected to get value.
	   printf("Buffered UpDown Counter is in progress...\nconnect the input signal to channels you selected to get value!\n");
	   printf("any key will stop Buffered UpDown Counter!\n\n");
	   while ( !kbhit()) {
         SLEEP(1);
      }

	   // step 8: Stop the operation if it is running.
	   ret= BufferedUdCounterCtrl_setEnabled(bfdUdCounterCtrl, FALSE);
	   CHK_RESULT(ret);
   }while(FALSE);

   // Step 9: Close device, release any allocated resource.
	BufferedUdCounterCtrl_Dispose(bfdUdCounterCtrl);

	// If something wrong in this execution, print the error code on screen for tracking.
   if(BioFailed(ret))
   {
      printf("\nSome error occurred. And the last error code is 0x%X.\n", ret);
      waitAnyKey();
   }  
   return 0;
}
